### ISSUE TITLE
Add set types, call it "species"?

### ISSUE THREAD

**Issue by jonsterling** [JonPRL/sml-red-jonprl#38]
Next up to be added is the Nuprl set type (i.e. set comprehension). I've been thinking a bit about the name (we have in the past used "subset"), and I am a bit unsatisfied with all the existing ones. "Set" is confusing because of Agda, "subset" is confusing because we don't call our types sets, and "subtype" is confusing because we might expect it to refer to the subtyping *relation*.

However, Brouwer used the word `Spezies` (English, *species*) for precisely a set that is defined by imposing a predicate on already-existing objects.

As far as the rules are concerned, this will be the first significant demonstration of the Dependent LCF machinery, in that we will be able to provide a single refinement rule for species-introduction.

````
species{σ} : (exp; [σ].exp)     for σ sort
```

**Comment by freebroccolo** [comment]
Think I missed this one. Are you set on the name species? I kind of wonder if it might cause some confusion with species in the sense of analytic functors. I think some of the French type theorists use "ensemble" which I kind of like but I don't know if that would be any clearer.

**Comment by jonsterling** [comment]
@freebroccolo I'm not set on the word species, and was also a little worried about the confusion with the combinatorial species etc... What do you prefer? I prefer "subtype comprehension" as the most clear term, but it is far too wordy. I don't like "subset" or "set"...

**Comment by freebroccolo** [comment]
Hmm, yeah… I'm not sure I have a clear choice. I agree the first one is the clearest but also agree that it is a bit long. What about just calling it "comprehension" and disambiguating in the documentation if necessary? As for ensemble, I believe this is the terminology Coq uses for sets in the style of predicates in their library. I'm not sure if that's a point in favor or against it… :) Anyway, I also agree that just calling it set or subset probably wouldn't be a good choice.

**Comment by jonsterling** [comment]
@freebroccolo I think "ensemble" sounds all right ("comprehension" is good, but it's longer)... In any case, I only need a name for the operator itself, since in practice, people will use the set comprehension notation.
